src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+
+                if (symbol == null)
+                {
+                    if (symbolItem.NavigationSymbolIndex < symbols.CandidateSymbols.Length)
+                    {
+                        symbol = symbols.CandidateSymbols[symbolItem.NavigationSymbolIndex.Value];
+                    }
+                    else
+                    {
+                        return null;
+                    }
+                }
+
+                var location = symbol.Locations.FirstOrDefault(l => l.SourceTree.Equals(tree));
+                return location ?? symbol.Locations.FirstOrDefault();
+            }
+
+            async Task<List<DocumentSymbol>> GetChildrenAsync(IEnumerable<NavigationBarItem> items)
+            {
+                var list = new List<DocumentSymbol>();
+                foreach (var item in items)
+                {
+                    list.Add(await CreateDocumentSymbolAsync(item).ConfigureAwait(false));
+                }
+
+                return list;
+            }
+
+            async Task<ISymbol> GetSymbolAsync(Location location)
+            {
+                var model = compilation.GetSemanticModel(location.SourceTree);
+                var root = await model.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+                var node = root.FindNode(location.SourceSpan);
+
+                while (node != null)
+                {
+                    var symbol = model.GetDeclaredSymbol(node);
+                    if (symbol != null)
+                    {
+                        return symbol;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Answers a workspace symbols request by providing a list of symbols found in a given workspace.
+        /// https://microsoft.github.io/language-server-protocol/specification#workspace_symbol
+        /// </summary>
+        /// <param name="solution">the current solution.</param>
+        /// <param name="request">the workspace request with the query to invoke.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>a list of symbols in the workspace.</returns>
+        public async Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(Solution solution, WorkspaceSymbolParams request, CancellationToken cancellationToken)
+        {
+            var symbols = ArrayBuilder<SymbolInformation>.GetInstance();
+
+            if (solution == null)
+            {
+                return symbols.ToArrayAndFree();
+            }
+
+            var searchTasks = solution.Projects.Select(
+                p => Task.Run(() => SearchProjectAsync(p), cancellationToken)).ToArray();
+
+            await Task.WhenAll(searchTasks).ConfigureAwait(false);
+
+            return symbols.ToArrayAndFree();
+
+            // local functions
+            async Task SearchProjectAsync(Project project)
```

Comment:
1. make this into a class.
2. if not '1', make this static and be clear about what is being passed alon.